d1985r3.md

```diff
@@ -222,6 +222,78 @@ OK, but and treating them like this also eases specification, and is more
 consistent with the way the language currently works. This is the more
 "adventurous" option, however.
 
+### Dependent names
+
+Universal template parameters are always dependent names. When a
+universal template parameter is used in an expression, it is parsed as a
+value by default. They can be disambiguated as a type by using `typename`
+as with other dependent names. To defer the type/value decision, `template
+auto` maybe used to perserve the name as a universal template. This allows
+the universal template parameter to be forwarded to another template.
+
+```cpp
+template<template auto Arg> void f()
+{
+  std::cout << Arg; // Arg is parsed a value
+  std::cout << sizeof(typename Arg); // Arg is parsed as a type
+
+  // Type/value decision is deferred to myclass
+  std::cout << myclass<template auto Arg>::value;
+}
+```
+
+The `typename auto` specifier may also be used to defer type/value lookup and
+form universal template parameters from other dependent names. This enables
+utility structs to perform transformations on packs of universal template
+parameters:
+
+
+```cpp
+template<typename T>
+struct unwrap_nttps;
+
+template<typename T>
+struct unwrap_nttps<T>
+{
+  using result = T;
+};
+
+template<typename T, T t>
+struct unwrap_nttps<std::integral_constant<T,t>>
+{
+  static inline constexpr T result = t;
+};
+```
+
+`template auto unwrap_nttps<Pack>::result...` may result in a template parameter
+pack of mixed types and NTTPs. By specifying the name of the dependent type as
+a universal template using `template auto`, the type/value state is
+deferred until it can be looked up during instantiation and a universal template
+parameter pack can be formed.
+
+A dependent name specified as a universal template parameter may not be used
+in any expression where the type/value information must be known for parsing:
+
+```cpp
+template<template auto... T>
+void f()
+{
+  std::cout << template auto T*2; // Error: Parse fails because T could be a type
+  myclass<template auto T*2> a; // Error: The same rule applies in template arguments.
```

Comment:
But what if myclass takes a UTP? No error, right?

Reply:
It's the `*2` part that would cause an error. A UTP with an operator applied to it doesn't parse.